### PR TITLE
Dynamic graph initial files and building

### PR DESCRIPTION
**Pull request by maawad**
The PR contains:

- Edge insertion for weighted directed and un-directed graphs
- Conversion from CSR/COO to dynamic graph in graph loader
- Unit tests for dynamic graph build and insertion

### CODE REVIEW

File: examples/dynamic_graph/test_dynamic_graph.cu [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    
+    using WeightedGraphT = app::TestGraph<VertexT, SizeT, ValueT,
+                                    graph::HAS_EDGE_VALUES | 
+                                    graph::HAS_CSR | 
+                                    graph::HAS_COO |
+                                    graph::HAS_DYN>;
+
+    using CSRGraphT = app::TestGraph<VertexT, SizeT, ValueT,
+                                    graph::HAS_EDGE_VALUES | 
+                                    graph::HAS_CSR>;
```
Comment (porumbes) [inline examples/dynamic_graph/test_dynamic_graph.cu:60]: Do we need `using CSRGraphT` for some reason?

Comment (maawad) [reply examples/dynamic_graph/test_dynamic_graph.cu:60]: I will remove this line. 
I was doing a complete conversion here but I moved everything to the google test file. This file will eventually contain an example to insertion/deletion.

Comment (maawad) [reply examples/dynamic_graph/test_dynamic_graph.cu]: Fixed in https://github.com/gunrock/gunrock/pull/713/commits/18776ca4aed356a2fd692983e006847e5496ebe1.

File: gunrock/graph/graph_base.cuh [hunk 2]
```diff
@@ -37,6 +37,7 @@ enum : GraphFlag {
   HAS_CSC = 0x0200,
   HAS_COO = 0x0400,
   HAS_GP = 0x0800,
+  HAS_DYN = 0x1000,
 
   GRAPH_PINNED = 0x1000,
```
Comment (porumbes) [inline gunrock/graph/graph_base.cuh:42]: Is this going to be an issue (same value for the enum)?

File: gunrock/graph/dynamic_graph/dynamic_graph.cuh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ */
+template<
+    typename _VertexT = int,
+    typename _SizeT   = _VertexT,
+    typename _ValueT  = _VertexT,
+    GraphFlag _FLAG   = GRAPH_NONE,
+    unsigned int cudaHostRegisterFlag = cudaHostRegisterDefault,
+    bool VALID = true,
+    bool HAS_VALUES = ((_FLAG & HAS_EDGE_VALUES) != 0)>
+struct Dyn : public DynamicGraphBase<_VertexT, _SizeT, _ValueT, _FLAG> {};
```
Comment (porumbes) [inline gunrock/graph/dynamic_graph/dynamic_graph.cuh:40]: `struct` defaults to public inheritance.

File: gunrock/graph/dynamic_graph/dynamic_graph_unweighted.cuh [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+ * @tparam ValueT Associated value type.
+ * @tparam GraphFlag graph flag
+ */
+template<
+    typename VertexT,
+    typename SizeT,
+    typename ValueT,
+    GraphFlag FLAG,
+    unsigned int cudaHostRegisterFlag>
+struct Dyn<VertexT, SizeT, ValueT, FLAG, cudaHostRegisterFlag, true, false> : public DynamicGraphBase<VertexT, SizeT, ValueT, FLAG> {
```
Comment (porumbes) [inline gunrock/graph/dynamic_graph/dynamic_graph_unweighted.cuh:38]: `struct` defaults to public inheritance.

File: gunrock/graph/graph_base.cuh (hunk 2 above)
Comment (maawad) [reply gunrock/graph/graph_base.cuh]: Nice catch. Fixed https://github.com/gunrock/gunrock/pull/713/commits/f937024a6415054821e33efb6d292d71ea4d323c.

File: gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+ * @tparam ValueT Associated value type.
+ * @tparam GraphFlag graph flag
+ */
+template<
+    typename VertexT,
+    typename SizeT,
+    typename ValueT,
+    GraphFlag FLAG,
+    unsigned int cudaHostRegisterFlag>
+struct Dyn<VertexT, SizeT, ValueT, FLAG, cudaHostRegisterFlag, true, true> : public DynamicGraphBase<VertexT, SizeT, ValueT, FLAG> {
```
Comment (porumbes) [inline gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh:37]: `struct` defaults to public inheritance.

Comment (porumbes) [inline gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh:37]: `struct` defaults to public inheritance.

File: gunrock/graph/graph_base.cuh [hunk 6]
```diff
@@ -37,6 +37,7 @@ enum : GraphFlag {
   HAS_CSC = 0x0200,
   HAS_COO = 0x0400,
   HAS_GP = 0x0800,
+  HAS_DYN = 0x2000,
 
   GRAPH_PINNED = 0x1000,
```
Comment (porumbes) [inline gunrock/graph/graph_base.cuh:42]: We might still have an issue here .. let's discuss.

File: gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh (hunk 5 above)
Comment (maawad) [reply gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh]: Fixed https://github.com/gunrock/gunrock/pull/713/commits/127da41c090f97105b46b91c36907068a040410c

**Pull request merged**